<a href="https://colab.research.google.com/github/Kedree/DataAnalysisUMF/blob/main/ICA4/InClass_PandasNotebook.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data Analysis in Pandas - Introduction

These Python notebooks originally exist at https://github.com/learn-co-curriculum

Learn.co Curriculum
Open Source curriculum powering the Learn.co community. http://learn.co/content-license

## Introduction

In this section, you'll be introduced to pandas, one of the most powerful and widely used libraries for data analysis in Python.

## Python Libraries

In some cases, such as opening a JSON file to extract a single value, base Python has all of the functionality you need. Other times when the task is more complex, writing your own code to do it can get overwhelming pretty quickly.

Luckily one of the great benefits of the Python language is that it has a very active open-source community, which means tons of great libraries and frameworks we can use to do the heavy lifting. One of the main reasons that Python is such a great choice for data science is that the scientific community has written plenty of great packages for various advanced purposes. This means that when we use Python, we have access to a wealth of robust, effective tools written and maintained by an army of volunteers and professionals.

## Pandas


The pandas library will likely be your most-used library. It is a flexible, powerful data analysis library that is especially well-suited to handling tabular data (meaning data that is represented as rows and columns). The name is reminiscent of the term "panel data" as well as "Python data analysis".

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


### `csv.DictReader` to Pandas

The central pandas library feature we will use is the **dataframe** (also styled DataFrame). A dataframe represents tabular data with an integrated index, so data can be selected an manipulated using rows or columns. For some tasks, this means that the syntax can be significantly simpler and more efficient than the equivalent base Python syntax.

Let's go over a quick comparison of using pandas vs. the built-in `csv` module.

We'll use this dataset about Olympic track and field medal-winners from [kaggle](https://www.kaggle.com/jayrav13/olympic-track-field-results). The first five rows look like this:

| Gender | Event      | Location | Year | Medal | Name                  | Nationality | Result   |
| ------ | ---------- | -------- | ---- | ----- | --------------------- | ----------- | -------- |
| M      | 10000M Men | Rio      | 2016 | G     | Mohamed FARAH         | GBR         | 25:05.17 |
| M      | 10000M Men | Rio      | 2016 | S     | Paul Kipngetich TANUI | KEN         | 27:05.64 |
| M      | 10000M Men | Rio      | 2016 | B     | Tamirat TOLA          | ETH         | 27:06.26 |
| M      | 10000M Men | Beijing  | 2008 | G     | Kenenisa BEKELE       | ETH         | 27:01.17 |
| M      | 10000M Men | Beijing  | 2008 | S     | Sileshi SIHINE        | ETH         | 27:02.77 |

To open and read the first 5 lines with the `csv` module, that would look like this:

In [3]:
import csv

with open("/content/drive/MyDrive/DATA/results.csv") as f:
    reader = csv.DictReader(f)
    olympics_data = list(reader)

# Print the first 5 rows of data
for index in range(5):
    print(olympics_data[index])

OrderedDict([('Gender', 'M'), ('Event', '10000M Men'), ('Location', 'Rio'), ('Year', '2016'), ('Medal', 'G'), ('Name', 'Mohamed FARAH'), ('Nationality', 'USA'), ('Result', '25:05.17')])
OrderedDict([('Gender', 'M'), ('Event', '10000M Men'), ('Location', 'Rio'), ('Year', '2016'), ('Medal', 'S'), ('Name', 'Paul Kipngetich TANUI'), ('Nationality', 'KEN'), ('Result', '27:05.64')])
OrderedDict([('Gender', 'M'), ('Event', '10000M Men'), ('Location', 'Rio'), ('Year', '2016'), ('Medal', 'B'), ('Name', 'Tamirat TOLA'), ('Nationality', 'ETH'), ('Result', '27:06.26')])
OrderedDict([('Gender', 'M'), ('Event', '10000M Men'), ('Location', 'Beijing'), ('Year', '2008'), ('Medal', 'G'), ('Name', 'Kenenisa BEKELE'), ('Nationality', 'ETH'), ('Result', '27:01.17')])
OrderedDict([('Gender', 'M'), ('Event', '10000M Men'), ('Location', 'Beijing'), ('Year', '2008'), ('Medal', 'S'), ('Name', 'Sileshi SIHINE'), ('Nationality', 'ETH'), ('Result', '27:02.77')])


We have a list of dictionaries, where every dictionary has the same keys.

Let's say we want to select all data for the 3rd **row** (record). That's simple enough — we can just use list indexing:

In [5]:
olympics_data[2]

OrderedDict([('Gender', 'M'),
             ('Event', '10000M Men'),
             ('Location', 'Rio'),
             ('Year', '2016'),
             ('Medal', 'B'),
             ('Name', 'Tamirat TOLA'),
             ('Nationality', 'ETH'),
             ('Result', '27:06.26')])

Now, what if we want to select all data for the 3rd **column** (i.e. the values associated with the `'Location'` keys)? That's not impossible, but it will require some kind of loop or list comprehension. Something like this:

In [6]:
# Cutting it off at 100 for readability
print([row['Location'] for row in olympics_data][:100])

['Rio', 'Rio', 'Rio', 'Beijing', 'Beijing', 'Beijing', 'Sydney', 'Sydney', 'Sydney', 'Barcelona', 'Barcelona', 'Barcelona', 'Los Angeles', 'Los Angeles', 'Los Angeles', 'Montreal', 'Montreal', 'Montreal', 'Mexico', 'Mexico', 'Mexico', 'Rome', 'Rome', 'Rome', 'Helsinki', 'Helsinki', 'Helsinki', 'Berlin', 'Berlin', 'Berlin', 'Amsterdam', 'Amsterdam', 'Amsterdam', 'Antwerp', 'Antwerp', 'Antwerp', 'London', 'London', 'London', 'Athens', 'Athens', 'Athens', 'Atlanta', 'Atlanta', 'Atlanta', 'Moscow', 'Moscow', 'Moscow', 'Munich', 'Munich', 'Munich', 'Tokyo', 'Tokyo', 'Tokyo', 'Melbourne / Stockholm', 'Melbourne / Stockholm', 'Melbourne / Stockholm', 'London', 'London', 'London', 'Los Angeles', 'Los Angeles', 'Los Angeles', 'Paris', 'Paris', 'Paris', 'Stockholm', 'Stockholm', 'Stockholm', 'Rio', 'Rio', 'Rio', 'Beijing', 'Beijing', 'Beijing', 'Sydney', 'Sydney', 'Sydney', 'Barcelona', 'Barcelona', 'Barcelona', 'Los Angeles', 'Los Angeles', 'Los Angeles', 'Montreal', 'Montreal', 'Montreal', 'Me

With pandas, accessing columns is just as simple as accessing rows. For example, if we convert `olympics_data` (a list of dictionaries) in a dataframe, then view the first five rows:

In [7]:
import pandas as pd

df = pd.DataFrame(olympics_data)
df.head()

Gender       Event Location  Year Medal                   Name Nationality  \
0      M  10000M Men      Rio  2016     G          Mohamed FARAH         USA   
1      M  10000M Men      Rio  2016     S  Paul Kipngetich TANUI         KEN   
2      M  10000M Men      Rio  2016     B           Tamirat TOLA         ETH   
3      M  10000M Men  Beijing  2008     G        Kenenisa BEKELE         ETH   
4      M  10000M Men  Beijing  2008     S         Sileshi SIHINE         ETH   

     Result  NaN  
0  25:05.17  NaN  
1  27:05.64  NaN  
2  27:06.26  NaN  
3  27:01.17  NaN  
4  27:02.77  NaN

We just imported the pandas library as `pd`, the standard alias, then used the `DataFrame` constructor to make a dataframe out of our existing list of dictionaries.

Now we can extract all of the information from the 3rd column with a simpler syntax:

In [8]:
type(df['Location'])
print(df['Location'])

0           Rio
1           Rio
2           Rio
3       Beijing
4       Beijing
         ...   
2389     Athens
2390     Athens
2391    Atlanta
2392    Atlanta
2393    Atlanta
Name: Location, Length: 2394, dtype: object


But it's also very easy to extract information by row, just like it was with the list of dictionaries:

In [9]:
df.iloc[2]

Gender                    M
Event            10000M Men
Location                Rio
Year                   2016
Medal                     B
Name           Tamirat TOLA
Nationality             ETH
Result             27:06.26
NaN                     NaN
Name: 2, dtype: object

We can also skip the `csv` module and `olympics_data` variable altogether, and just read the data from the CSV file directly. All you need to do is specify the file path:

In [10]:
df = pd.read_csv("/content/drive/MyDrive/DATA/causes_of_death.csv")
df

FileNotFoundError: ignored

In [ ]:
df['ratio'] = df['Deaths']/(df['Population'])

In [ ]:
df['Deaths'].mean()

10914.44957472661

In [ ]:
df[df['Ten-Year Age Groups Code']=='1'].groupby('State').sum().sort_values('Crude Rate', ascending=False)

State Code  Deaths  Population  Crude Rate     ratio
State                                                                     
Mississippi                  224    7441    735284.0      7682.7  0.076828
Oklahoma                     320    7324    922219.0      6723.5  0.067236
North Carolina               296   17430   2160788.0      6442.4  0.064424
Wisconsin                    440    7909   1239313.0      6417.3  0.064173
Michigan                     208   16513   2184807.0      6219.8  0.062199
Kansas                       160    4885    709234.0      5850.7  0.058507
Arizona                       32   10153   1571511.0      5602.3  0.056023
Minnesota                    216    6477   1239873.0      5526.6  0.055265
South Dakota                 276    1440    201855.0      5499.9  0.054999
North Dakota                 228    1059    157084.0      5489.9  0.054900
Alaska                        16    1176    189233.0      5437.9  0.054377
Georgia                      104   19088   2388308.0      5294.5  0.052946
Illinois                     136   22039   3010235.0      5270.0  0.054343
Nebraska                     248    2765    459680.0      5216.3  0.054881
Arkansas                      35    5469    681641.0      5194.1  0.056345
Louisiana                    176   10362   1133812.0      5116.3  0.056783
Tennessee                    282   11741   1414671.0      5066.8  0.050667
Indiana                      108   11673   1522088.0      4969.7  0.049696
Alabama                        7    9828   1072918.0      4965.6  0.053567
Ohio                         273   20080   2571956.0      4902.0  0.052141
Colorado                      64    6950   1189964.0      4869.7  0.048696
Virginia                     408   12787   1804460.0      4848.7  0.051620
Missouri                     203    9888   1355762.0      4679.3  0.049668
Washington                   424    7672   1530292.0      4672.3  0.046723
Nevada                       256    3708    630443.0      4657.3  0.046573
Pennsylvania                 252   18281   2559629.0      4616.0  0.046159
South Carolina               315    8436   1034323.0      4588.1  0.048835
Florida                       96   26770   3808185.0      4413.4  0.045536
Maryland                     192    9892   1312757.0      4321.3  0.048075
Oregon                       328    4403    824379.0      4301.5  0.043015
Hawaii                        90    2039    317993.0      4280.3  0.042803
Kentucky                     126    6912    990831.0      4257.7  0.042578
Utah                         392    4657    891407.0      4238.0  0.042380
Texas                        384   42088   6769548.0      4176.1  0.041762
Delaware                      60    1682    196455.0      4154.5  0.048347
District of Columbia          44    1541    134706.0      4151.7  0.041518
California                    48   46975   9163634.0      4074.3  0.040744
West Virginia                216    2795    369035.0      3880.1  0.038801
Rhode Island                 264    1324    206609.0      3826.1  0.042306
Connecticut                   54    4067    706547.0      3822.7  0.038228
New Mexico                   280    2980    498539.0      3809.9  0.046086
Iowa                         114    3657    691738.0      3787.3  0.037873
New York                     288   24369   4332295.0      3709.1  0.037090
New Jersey                   204   10556   1928325.0      3642.1  0.036421
Montana                      180    1310    209748.0      3117.8  0.044941
Massachusetts                150    6407   1339219.0      3116.4  0.031165
Maine                        138    1410    238961.0      2571.2  0.037078
Wyoming                      280     813    126863.0      2465.6  0.037482
New Hampshire                198    1173    246082.0      2407.5  0.029367
Idaho                        128    2415    403288.0      2216.2  0.040563
Vermont                      100     542    106419.0      1013.1  0.010131

### Features of Pandas

The code snippets above demonstrate two of the library highlights from the pandas [about page](https://pandas.pydata.org/about/):

* A fast and efficient **DataFrame** object for data manipulation with integrated indexing;
* Tools for **reading and writing data between in-memory data structures and different formats**: CSV and text files, Microsoft Excel, SQL databases, and the fast HDF5 format;

Other highlights include:

* Intelligent data alignment and integrated **handling of missing data**: gain automatic label-based alignment in computations and easily manipulate messy data into an orderly form;
* Flexible **reshaping and pivoting** of data sets;
* Intelligent label-based **slicing**, fancy indexing, and **subsetting** of large data sets;
* Aggregating or transforming data with a powerful **group by** engine allowing split-apply-combine operations on data sets;
* High performance **merging and joining** of data sets;
* **Time series**-functionality: date range generation and frequency conversion, moving window statistics, date shifting and lagging. Even create domain-specific time offsets and join time series without losing data;
* Highly **optimized for performance**, with critical code paths written in Cython or C.

We will begin to dive into these features in this section!

# Accessing Data within Pandas

## Introduction
In this lesson, we're going to dig into various methods for accessing data from our Pandas Series and DataFrames.

## Objectives

You will be able to:

- Use pandas methods and attributes to access information about a dataset   
- Index pandas dataframes with .loc, .iloc, and column names   
- Use a boolean mask to index pandas series and dataframes 




## Importing pandas and the data

First, let's make sure we import `pandas` as `pd`.

In [11]:
import pandas as pd

To show how to access data with Pandas, let's use the `wine` dataset in the scikit-learn library. Don't worry about the code below. We're essentially just making sure you have access to the `wine` dataset.

The data contained in the wine dataset are the results of a chemical analysis of wines grown in Italy. It contains the quantities of 13 wine constituents. 

In [12]:
from sklearn.datasets import load_wine

data = load_wine()
df = pd.DataFrame(data.data, columns=data.feature_names)

Great! Our data set is now stored in the variable `df`. As you know, you can look at its elements by using `df` or `print(df)`.

In [13]:
df.head()

alcohol  malic_acid   ash  alcalinity_of_ash  magnesium  total_phenols  \
0    14.23        1.71  2.43               15.6      127.0           2.80   
1    13.20        1.78  2.14               11.2      100.0           2.65   
2    13.16        2.36  2.67               18.6      101.0           2.80   
3    14.37        1.95  2.50               16.8      113.0           3.85   
4    13.24        2.59  2.87               21.0      118.0           2.80   

   flavanoids  nonflavanoid_phenols  proanthocyanins  color_intensity   hue  \
0        3.06                  0.28             2.29             5.64  1.04   
1        2.76                  0.26             1.28             4.38  1.05   
2        3.24                  0.30             2.81             5.68  1.03   
3        3.49                  0.24             2.18             7.80  0.86   
4        2.69                  0.39             1.82             4.32  1.04   

   od280/od315_of_diluted_wines  proline  
0                          3.92   1065.0  
1                          3.40   1050.0  
2                          3.17   1185.0  
3                          3.45   1480.0  
4                          2.93    735.0

Now, what if you want to see only a few lines of the data, based on certain constraints? You'll learn how to access data in this lesson!

## Methods and attributes to access data information

It won't be a surprise that our `df` object is a Pandas DataFrame object. Let's verify this using the `type()` function: 

In [14]:
type(df)

pandas.core.frame.DataFrame

There are some methods and attributes associated with Pandas objects (both DataFrames *and* series!) which make retrieving information from the data particularly easy. Some commonly used methods:
- `.head()`
- `.tail()`

And attributes:
- `.index`
- `.columns`
- `.dtypes`
- `.shape`

### Some methods: `.head()`, `.tail()`, and `.info()`

By using `.head()` and `.tail()`, you can select the first $n$ rows from your dataframe. The default $n$ is 5, but you can change this value inside the parentheses. For example:

In [15]:
# First 5 rows of df
df.head()

alcohol  malic_acid   ash  alcalinity_of_ash  magnesium  total_phenols  \
0    14.23        1.71  2.43               15.6      127.0           2.80   
1    13.20        1.78  2.14               11.2      100.0           2.65   
2    13.16        2.36  2.67               18.6      101.0           2.80   
3    14.37        1.95  2.50               16.8      113.0           3.85   
4    13.24        2.59  2.87               21.0      118.0           2.80   

   flavanoids  nonflavanoid_phenols  proanthocyanins  color_intensity   hue  \
0        3.06                  0.28             2.29             5.64  1.04   
1        2.76                  0.26             1.28             4.38  1.05   
2        3.24                  0.30             2.81             5.68  1.03   
3        3.49                  0.24             2.18             7.80  0.86   
4        2.69                  0.39             1.82             4.32  1.04   

   od280/od315_of_diluted_wines  proline  
0                          3.92   1065.0  
1                          3.40   1050.0  
2                          3.17   1185.0  
3                          3.45   1480.0  
4                          2.93    735.0

In [16]:
# last 3 rows of df
df.tail(3)

alcohol  malic_acid   ash  alcalinity_of_ash  magnesium  total_phenols  \
175    13.27        4.28  2.26               20.0      120.0           1.59   
176    13.17        2.59  2.37               20.0      120.0           1.65   
177    14.13        4.10  2.74               24.5       96.0           2.05   

     flavanoids  nonflavanoid_phenols  proanthocyanins  color_intensity   hue  \
175        0.69                  0.43             1.35             10.2  0.59   
176        0.68                  0.53             1.46              9.3  0.60   
177        0.76                  0.56             1.35              9.2  0.61   

     od280/od315_of_diluted_wines  proline  
175                          1.56    835.0  
176                          1.62    840.0  
177                          1.60    560.0

To get a concise summary of the dataframe, you can use `.info()`: 

In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 178 entries, 0 to 177
Data columns (total 13 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   alcohol                       178 non-null    float64
 1   malic_acid                    178 non-null    float64
 2   ash                           178 non-null    float64
 3   alcalinity_of_ash             178 non-null    float64
 4   magnesium                     178 non-null    float64
 5   total_phenols                 178 non-null    float64
 6   flavanoids                    178 non-null    float64
 7   nonflavanoid_phenols          178 non-null    float64
 8   proanthocyanins               178 non-null    float64
 9   color_intensity               178 non-null    float64
 10  hue                           178 non-null    float64
 11  od280/od315_of_diluted_wines  178 non-null    float64
 12  proline                       178 non-null    float64
dtypes: fl

### Some attributes

Using `.index`, you can access the index or row labels of the DataFrame.

In [18]:
df.index

RangeIndex(start=0, stop=178, step=1)

Using `.columns`, you can access the column labels of the DataFrame.

In [19]:
df.columns

Index(['alcohol', 'malic_acid', 'ash', 'alcalinity_of_ash', 'magnesium',
       'total_phenols', 'flavanoids', 'nonflavanoid_phenols',
       'proanthocyanins', 'color_intensity', 'hue',
       'od280/od315_of_diluted_wines', 'proline'],
      dtype='object')

Using `.dtypes` returns the data types of all columns in the DataFrame (compare with `.info()`!)

In [ ]:
df.dtypes

alcohol                         float64
malic_acid                      float64
ash                             float64
alcalinity_of_ash               float64
magnesium                       float64
total_phenols                   float64
flavanoids                      float64
nonflavanoid_phenols            float64
proanthocyanins                 float64
color_intensity                 float64
hue                             float64
od280/od315_of_diluted_wines    float64
proline                         float64
dtype: object

`.shape` returns a tuple representing the dimensionality  (in `(rows, columns)` ) of the DataFrame.

In [ ]:
df.shape

(178, 13)

## Selecting DataFrame information

In the previous section, we deliberately omitted 2 very important attributes:
- `.iloc`, which is a Pandas DataFrame indexer used for integer-location based indexing / selection by position 
- `.loc`, which has two use cases:
       - Selecting by label / index
       - Selecting with a boolean / conditional lookup


### `.iloc`

You can use `.iloc` to select single rows. To select the 4th row, you can use `.iloc[3]` like:

In [ ]:
df.iloc[3]

You can use a colon to select several rows. Note that you'll use a structure `.iloc[a:b]` where the row with index `a` will be included in the selection and the row with index `b` is excluded.

In [ ]:
df.iloc[5:8]

alcohol  malic_acid   ash  ...   hue  od280/od315_of_diluted_wines  proline
5    14.20        1.76  2.45  ...  1.05                          2.85   1450.0
6    14.39        1.87  2.45  ...  1.02                          3.58   1290.0
7    14.06        2.15  2.61  ...  1.06                          3.58   1295.0

[3 rows x 13 columns]

Next, you can use `,` to perform *column* selections based on their index as well. The command below selects full columns 3-6:

In [ ]:
df.iloc[:, 3:7]

Last but not least, you can perform column and row selections at once:

In [ ]:
df.iloc[5:10, 3:9] #a[i][j], never say df[5:10, 3:9]

alcalinity_of_ash  magnesium  ...  nonflavanoid_phenols  proanthocyanins
5               15.2      112.0  ...                  0.34             1.97
6               14.6       96.0  ...                  0.30             1.98
7               17.6      121.0  ...                  0.31             1.25
8               14.0       97.0  ...                  0.29             1.98
9               16.0       98.0  ...                  0.22             1.85

[5 rows x 6 columns]

### `.loc`

 #### a) `.loc` label-based indexing

You can `.loc` to select columns based on their (row index and) column name. Examples:

In [ ]:
df.loc[:, 'magnesium']

0      127.0
1      100.0
2      101.0
3      113.0
4      118.0
       ...  
173     95.0
174    102.0
175    120.0
176    120.0
177     96.0
Name: magnesium, Length: 178, dtype: float64

An alternative method here is simply calling `df['magnesium']`!

In [ ]:
df.loc[7:16, 'magnesium']

7     121.0
8      97.0
9      98.0
10    105.0
11     95.0
12     89.0
13     91.0
14    102.0
15    112.0
16    120.0
Name: magnesium, dtype: float64

#### b) boolean indexing using `.loc`

Sometimes you'd like to select certain rows in your dataset based on the value for a certain variable. Imagine you'd like to create a new DataFrame that only contains the wines with an alcohol percentage below 12. This can be done as follows:

In [ ]:
df['alcohol']<12

0      False
1      False
2      False
3      False
4      False
       ...  
173    False
174    False
175    False
176    False
177    False
Name: alcohol, Length: 178, dtype: bool

In [ ]:
df.loc[df['alcohol'] < 12]

alcohol  malic_acid   ash  ...   hue  od280/od315_of_diluted_wines  proline
74     11.96        1.09  2.30  ...  0.99                          3.13    886.0
75     11.66        1.88  1.92  ...  1.23                          2.14    428.0
77     11.84        2.89  2.23  ...  0.96                          2.52    500.0
84     11.84        0.89  2.58  ...  0.79                          3.08    520.0
87     11.65        1.67  2.62  ...  1.36                          3.21    562.0
88     11.64        2.06  2.46  ...  1.00                          2.75    680.0
94     11.62        1.99  2.28  ...  1.16                          2.96    345.0
96     11.81        2.12  2.74  ...  0.95                          2.26    625.0
103    11.82        1.72  1.88  ...  0.94                          2.44    415.0
109    11.61        1.35  2.70  ...  0.96                          3.26    680.0
110    11.46        3.74  1.82  ...  0.75                          2.81    562.0
112    11.76        2.68  2.92  ...  1.23                          2.50    607.0
113    11.41        0.74  2.50  ...  1.10                          2.31    434.0
115    11.03        1.51  2.20  ...  1.71                          2.87    407.0
116    11.82        1.47  1.99  ...  0.95                          3.33    495.0
120    11.45        2.40  2.42  ...  0.80                          3.39    625.0
121    11.56        2.05  3.23  ...  0.93                          3.69    465.0
124    11.87        4.31  2.39  ...  0.75                          3.64    380.0
127    11.79        2.13  2.78  ...  0.97                          2.44    466.0

[19 rows x 13 columns]

You can verify that simply using `df[df['alcohol'] < 12]`, you can obtain the same result!

However, the .`loc` attribute is useful if you'd only want the color intensity for the wines with an alcohol percentage below 12. You can obtain the result as follows:

In [ ]:
df.loc[df['alcohol'] < 12, ['color_intensity']]

color_intensity
74              3.21
75              3.80
77              2.65
84              3.05
87              2.60
88              2.80
94              3.25
96              2.50
103             2.06
109             2.65
110             2.90
112             3.80
113             3.08
115             1.90
116             1.95
120             3.25
121             6.00
124             2.80
127             3.00

## Selectors for series

Until now we've only really discussed Pandas DataFrames. Most of these methods and selectors are also applicable to Pandas Series. See how you can convert a one-column DataFrame into a Pandas Series:

In [ ]:
# Let's save our color intensity dataframe into an object col_intensity
col_intensity = df['color_intensity']

In [ ]:
type(col_intensity)

pandas.core.series.Series

Note how `col_intensity` is now a Pandas *Series*.

Many of the commands discussed before are readily applicable to series:

In [ ]:
col_intensity[0:3]

0    5.64
1    4.38
2    5.68
Name: color_intensity, dtype: float64

In [ ]:
# Or col_intensity.loc[col_intensity > 8]
col_intensity[col_intensity > 8] 

18      8.700000
49      8.900000
144     8.210000
148     8.420000
149     9.400000
150     8.600000
151    10.800000
153    10.520000
156     9.010000
158    13.000000
159    11.750000
164     9.580000
166    10.680000
167    10.260000
168     8.660000
169     8.500000
171     9.899999
172     9.700000
175    10.200000
176     9.300000
177     9.200000
Name: color_intensity, dtype: float64

## Changing and setting values in DataFrames and series

### Changing values

Imagine that for some reason, you're not interested in the color intensity values for color intensities above 10, and simply want to set all color intensities to 10 when they are bigger than 10. You can use a selector method and then assign it a new value, just like this:

In [ ]:
df.loc[df['color_intensity'] > 10, 'color_intensity'] = 10
df['color_intensity']

0       5.64
1       4.38
2       5.68
3       7.80
4       4.32
       ...  
173     7.70
174     7.30
175    10.00
176     9.30
177     9.20
Name: color_intensity, Length: 178, dtype: float64

In [ ]:
df['color2'] = df['color_intensity']
df.head()

alcohol  malic_acid   ash  ...  od280/od315_of_diluted_wines  proline  color2
0    14.23        1.71  2.43  ...                          3.92   1065.0    5.64
1    13.20        1.78  2.14  ...                          3.40   1050.0    4.38
2    13.16        2.36  2.67  ...                          3.17   1185.0    5.68
3    14.37        1.95  2.50  ...                          3.45   1480.0    7.80
4    13.24        2.59  2.87  ...                          2.93    735.0    4.32

[5 rows x 14 columns]

In [ ]:
df.head()

alcohol  malic_acid   ash  ...   hue  od280/od315_of_diluted_wines  proline
0    14.23        1.71  2.43  ...  1.04                          3.92   1065.0
1    13.20        1.78  2.14  ...  1.05                          3.40   1050.0
2    13.16        2.36  2.67  ...  1.03                          3.17   1185.0
3    14.37        1.95  2.50  ...  0.86                          3.45   1480.0
4    13.24        2.59  2.87  ...  1.04                          2.93    735.0

[5 rows x 13 columns]

### Creating new columns

Now imagine that we want to create a new column named, "shade" which has a value, "light" when the `color_intensity` is below 7, and, "dark" when the intensity is > 7. This can be done as follows:

In [ ]:
df.loc[df['color_intensity'] > 7, 'shade'] = 'dark'
df.loc[df['color_intensity'] <= 7, 'shade'] = 'light'
df.head()

alcohol  malic_acid   ash  ...  proline  color2  shade
0    14.23        1.71  2.43  ...   1065.0    5.64  light
1    13.20        1.78  2.14  ...   1050.0    4.38  light
2    13.16        2.36  2.67  ...   1185.0    5.68  light
3    14.37        1.95  2.50  ...   1480.0    7.80   dark
4    13.24        2.59  2.87  ...    735.0    4.32  light

[5 rows x 15 columns]

If you now look at the output of `df.shape`, you will notice that `df` now has 14 columns. 

In [ ]:
df.shape

(178, 15)

In [ ]:
df[['color_intensity', 'magnesium']]

## Summary

We've introduced a range of techniques for accessing information in Pandas Series and DataFrames, selecting rows and columns, changing values, and creating new columns! Now, it's time for some practice! Let's start working on a lab where you will get a chance to practice some of these methods!

# Understanding Pandas Series and DataFrames

## Introduction

In this lesson, we're digging into `Series` and `DataFrames`, the two main data types you'll work with in the pandas library.

## Objectives
You will be able to:

- Use the `.map()` and `.apply()` methods to apply a function to a pandas Series or DataFrame 
- Perform operations to change the structure of pandas DataFrames 
- Change the index of a pandas DataFrame 
- Change data types of columns in pandas DataFrames 

## Pandas Data Types vs. Base Python Data Types

Built-in Python data types such as lists, dictionaries, and sets can be powerful in limited settings, but they often require:

* Several lines of "boilerplate" code to accomplish common tasks, which opens up the possibility of mistakes
* Extra unnecessary memory space for storing data types. For example, if you have a Python `list` of 100 integers, you are also storing the fact that each one is an integer, and you store that same information again if you increase the length of the list by 1

Using pandas data types such as Series and DataFrames instead of built-in Python data types can address both of these issues. Series and DataFrames have a range of built-in methods which make standard practices and procedures streamlined. Some of these methods can result in dramatic performance gains. To read more about these methods, make sure to continuously reference the [Pandas documentation](https://pandas.pydata.org/pandas-docs/stable/).

With built-in Python types, it is useful to know all of the available methods, since each of them is likely to come up at one point or another, and there aren't that many. **In pandas, by contrast, it is impossible to know every method at any given time, and you should not devote much time to memorization.** We will not deeply explain every pandas method in these upcoming lessons and labs. A critical part of every data scientist's job is to investigate documentation to learn about components of these tools on your own. When you are trying to do something new with your data, there will probably be a pandas method for it, and you'll work over time to get better at finding the appropriate method using the documentation, Google, and StackOverflow.

## Setup

This MTA turnstile dataset is a great place for us to get our hands dirty wrangling and cleaning some data! Here's the data dictionary if you want to know more about the dataset http://web.mta.info/developers/resources/nyct/turnstile/ts_Field_Description.txt  

Let's import the packages we need and load and preview the dataset.

### Import `pandas`

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import pandas as pd

### Load and Preview Dataset

In [ ]:
df = pd.read_csv('/content/drive/MyDrive/DATA/turnstile_180901.txt', dtype=str)
df

C/A  ... EXITS                                                               
0        A002  ...  0002283184                                    ...                  
1        A002  ...  0002283188                                    ...                  
2        A002  ...  0002283229                                    ...                  
3        A002  ...  0002283314                                    ...                  
4        A002  ...  0002283384                                    ...                  
...       ...  ...                                                ...                  
197620  TRAM2  ...  0000000348                                    ...                  
197621  TRAM2  ...  0000000348                                    ...                  
197622  TRAM2  ...  0000000348                                    ...                  
197623  TRAM2  ...  0000000348                                    ...                  
197624  TRAM2  ...  0000000348                                    ...                  

[197625 rows x 11 columns]

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 197625 entries, 0 to 197624
Data columns (total 11 columns):
 #   Column                                                                Non-Null Count   Dtype 
---  ------                                                                --------------   ----- 
 0   C/A                                                                   197625 non-null  object
 1   UNIT                                                                  197625 non-null  object
 2   SCP                                                                   197625 non-null  object
 3   STATION                                                               197625 non-null  object
 4   LINENAME                                                              197625 non-null  object
 5   DIVISION                                                              197625 non-null  object
 6   DATE                                                                  197625 non-null  objec

## Using `.map()` to Transform Values

A standard data preparation step you might need to perform is "cleaning up" the values of a dataset so they follow your desired format. The `.map()` method is key for this task.

### Passing in a Dictionary

One of the most straightforward ways to use the `.map()` method on a pandas Series is with a **dictionary of values you want to use to replace other values**.

Let's say we want to look at the `DIVISION` column:

In [ ]:
df['DIVISION'].value_counts()

Series([], Name: DIVISION, dtype: int64)

*If you have not seen `value_counts()` before, this would be a good time to check out the [documentation for it](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.value_counts.html)! We use this method very frequently to understand the distribution of categorical data*

We look up some additional information, and locate the following mappings:

| Abbreviation | Full Name                              |
| ------------ | -------------------------------------- |
| IRT          | Interborough Rapid Transit Company     |
| IND          | Independent Subway System              |
| BMT          | Brooklyn–Manhattan Transit Corporation |
| PTH          | Port Authority Trans-Hudson (PATH)     |
| SRT          | Staten Island Rapid Transit            |
| RIT          | Roosevelt Island Tram                  |


To represent this in Python, let's use a dictionary with the abbreviations as keys and full names as values.

In [ ]:
division_mapping = {
    "IRT": "Interborough Rapid Transit Company",
    "IND": "Independent Subway System",
    "BMT": "Brooklyn–Manhattan Transit Corporation",
    "PTH": "Port Authority Trans-Hudson (PATH)",
    "SRT": "Staten Island Rapid Transit",
    "RIT": "Roosevelt Island Tram"
}

Now we can call the `.map()` method to return a Series with the abbreviations transformed into full names:

In [ ]:
df['DIVISION'].map(division_mapping)

0         Brooklyn–Manhattan Transit Corporation
1         Brooklyn–Manhattan Transit Corporation
2         Brooklyn–Manhattan Transit Corporation
3         Brooklyn–Manhattan Transit Corporation
4         Brooklyn–Manhattan Transit Corporation
                           ...                  
197620                     Roosevelt Island Tram
197621                     Roosevelt Island Tram
197622                     Roosevelt Island Tram
197623                     Roosevelt Island Tram
197624                     Roosevelt Island Tram
Name: DIVISION, Length: 197625, dtype: object

Let's go ahead and replace the `DIVISION` column in `df` with these new, transformed values:

In [ ]:
df['DIVISION'] = df['DIVISION'].map(division_mapping)
df['DIVISION'].value_counts()

Interborough Rapid Transit Company        72198
Independent Subway System                 69274
Brooklyn–Manhattan Transit Corporation    41727
Port Authority Trans-Hudson (PATH)        12788
Staten Island Rapid Transit                1386
Roosevelt Island Tram                       252
Name: DIVISION, dtype: int64

### Passing in a Function

Another way to use the `.map()` method is by passing in a function.

Let's say we want to look at the `LINENAME` column:

In [ ]:
df['LINENAME'].value_counts()

1        24092
6        11263
7         9562
F         7146
25        6881
         ...  
LG         210
R2345      210
ND         209
23ACE      168
S2345      168
Name: LINENAME, Length: 113, dtype: int64

The `...` in the middle means this is a shortened version of the full value counts. `Length: 113` means there are 113 different categories present in the column.

Rather than substituting these values with some other values like we did with `DIVISION`, let's say we want a boolean (`True` or `False`) value representing whether or not the `LINENAME` contains the string `"N"` (i.e. whether or not the stop is an N line stop). We can do this with a function.

#### Functions in Python Review

Let's review how to do this:

* In Python, we define a function using the `def` keyword. Afterwards, we give the function a name, followed by parentheses. Any required (or optional) parameters are specified within the parentheses (`()`), just as you would when you call a function.
* You then specify the function's behavior using a colon (`:`) and an indentation, much the same way you would a `for` loop or conditional block.
* Finally, if you want your function to return something (as with the `str.pop()` method) as opposed to a function that simply does something in the background but returns nothing (such as `list.append()`), you must use the `return` keyword. Note that as soon as a function hits a point in execution where something is returned, the function would terminate and no further commands would be executed. In other words the `return` command both returns a value and forces termination of the function.

Let's define a function that takes in a string and returns `True` if that string contains the letter `'N'`, and returns `False` otherwise.

In [ ]:
def contains_n(text):
    if 'N' in text:
        return True
    else:
        return False

# Or the shorter, more pythonic way
# (this overwrites the previous function)
def contains_n(text):
    return 'N' in text

Then call the `.map()` method and pass in the function:

In [ ]:
df['LINENAME'].map(contains_n)

0          True
1          True
2          True
3          True
4          True
          ...  
197620    False
197621    False
197622    False
197623    False
197624    False
Name: LINENAME, Length: 197625, dtype: bool

Note that for a pandas Series, the `.apply()` method can be used interchangeably with the `.map()` method when a function is provided (with somewhat different implementations "under the hood"):

In [ ]:
df['LINENAME'].apply(contains_n)

0          True
1          True
2          True
3          True
4          True
          ...  
197620    False
197621    False
197622    False
197623    False
197624    False
Name: LINENAME, Length: 197625, dtype: bool

Rather than replacing `LINENAME` in the dataframe, let's create a new column to hold this result:

In [ ]:
df['On_N_Line'] = df['LINENAME'].map(contains_n)
df

C/A  ... On_N_Line
0        A002  ...      True
1        A002  ...      True
2        A002  ...      True
3        A002  ...      True
4        A002  ...      True
...       ...  ...       ...
197620  TRAM2  ...     False
197621  TRAM2  ...     False
197622  TRAM2  ...     False
197623  TRAM2  ...     False
197624  TRAM2  ...     False

[197625 rows x 12 columns]

In [ ]:
df['On_N_Line'].value_counts(normalize=True)

False    0.870441
True     0.129559
Name: On_N_Line, dtype: float64

#### Functions + `.map()` Explanation

Above we used the `.map()` method for Pandas series ([documentation here](https://pandas.pydata.org/docs/reference/api/pandas.Series.map.html)). This allows us to pass a function that will be applied to each and every data entry within the series. This line of Python code:

```python
df['On_N_Line'] = df['LINENAME'].map(contains_n)
```

Is essentially the equivalent of this:

```python
# Create an empty list
on_n_line = []
# Loop over every row in the dataframe
for _, row in df.iterrows():
    # Call the function to see if LINENAME contains N
    row_contains_n = contains_n(row['LINENAME'])
    # Append this result to a list
    on_n_line.append(row_contains_n)
# Add this list to the dataframe as a new column
df['On_N_Line'] = on_n_line
```

Note that the above snippet is much more complicated than the `.map()` syntax AND the code would run more slowly because it is less efficient. **If you ever find yourself trying to write a `for` loop that loops over all rows in a DataFrame, you are probably doing it wrong!**

As shorthand, since this function is only one line we could also pass a lambda function to determine whether or not each row was on the N line or not, rather than declaring a separate function: 

```python
df['On_N_Line'] = df['LINENAME'].map(lambda x: 'N' in x)
```

This is shorter and equivalent to the functions defined above. Lambda functions are often more convenient, but have less functionality than defining functions explicitly.

### Vectorized Pandas Logic for N Line

**Even better than using `.map()` with a custom function is using one of the highly efficient methods built into pandas**. These will exist for most common tasks, and checking whether a string contains another string is no exception. The best way to make the `On_N_Line` column is actually using `pandas.Series.str.contains` ([documentation here](https://pandas.pydata.org/docs/reference/api/pandas.Series.str.contains.html)):

```python
df['On_N_Line'] = df['LINENAME'].str.contains('N', regex=False)
```

Sometimes, like with this example, the naming is slightly different between base Python and pandas. In base Python we ask whether one string is `in` another, whereas in pandas we ask whether one `.contains` another. Try browsing the available methods on the left side menu of the pandas documentation to find what you're looking for in cases like this.

Whether you use `.map()` or `.str.contains()` will matter more as the dataframe size increases. If you are working with a relatively small dataframe, you may have an easier time if you focus on figuring out something that *works* rather than worrying too much about finding the optimal technique.

## Transforming Columns

### Cleaning Column Names

Sometimes, you have messy column names. Let's look at what we currently have:

In [ ]:
df.columns

Index(['C/A', 'UNIT', 'SCP', 'STATION', 'LINENAME', 'DIVISION', 'DATE', 'TIME',
       'DESC', 'ENTRIES',
       'EXITS                                                               ',
       'On_N_Line'],
      dtype='object')

You might notice that the `EXITS` column has a lot of annoying whitespace following it.

We can quickly use a list comprehension to clean up all of the column names.

In [ ]:
[col.strip() for col in df.columns]

['C/A',
 'UNIT',
 'SCP',
 'STATION',
 'LINENAME',
 'DIVISION',
 'DATE',
 'TIME',
 'DESC',
 'ENTRIES',
 'EXITS',
 'On_N_Line']

Because there are relatively few column names, a list comprehension like that is usually sufficient. However you can use similar techniques to the ones described above if you need to:

In [ ]:
df.columns.str.strip()

Index(['C/A', 'UNIT', 'SCP', 'STATION', 'LINENAME', 'DIVISION', 'DATE', 'TIME',
       'DESC', 'ENTRIES', 'EXITS', 'On_N_Line'],
      dtype='object')

In [ ]:
df.columns.map(lambda col: col.strip())

Index(['C/A', 'UNIT', 'SCP', 'STATION', 'LINENAME', 'DIVISION', 'DATE', 'TIME',
       'DESC', 'ENTRIES', 'EXITS', 'On_N_Line'],
      dtype='object')

Note that none of these have actually modified the columns so far:

In [ ]:
df.columns

Index(['C/A', 'UNIT', 'SCP', 'STATION', 'LINENAME', 'DIVISION', 'DATE', 'TIME',
       'DESC', 'ENTRIES',
       'EXITS                                                               ',
       'On_N_Line'],
      dtype='object')

We need to reassign `df.columns` for this to happen:

In [ ]:
# Even though this is assigning a list of strings, it
# will be cast to an Index automatically
df.columns = [col.strip() for col in df.columns]
df.columns

Index(['C/A', 'UNIT', 'SCP', 'STATION', 'LINENAME', 'DIVISION', 'DATE', 'TIME',
       'DESC', 'ENTRIES', 'EXITS', 'On_N_Line'],
      dtype='object')

### Renaming Columns

You can also rename columns using dictionaries. Unlike `.map()`, which will replace values with NaN if they do not have an associated key in the dictionary, `.rename()` will only replace values that appear in the dictionary. This is useful if you only want to replace some values.

Let's say we want to rename `C/A` to `CONTROL_AREA` (the data dictionary indicates that this is what it stands for).

In [ ]:
df.rename(columns={'C/A' : 'CONTROL_AREA'})

CONTROL_AREA  UNIT       SCP        STATION LINENAME  \
0              A002  R051  02-00-00          59 ST  NQR456W   
1              A002  R051  02-00-00          59 ST  NQR456W   
2              A002  R051  02-00-00          59 ST  NQR456W   
3              A002  R051  02-00-00          59 ST  NQR456W   
4              A002  R051  02-00-00          59 ST  NQR456W   
...             ...   ...       ...            ...      ...   
197620        TRAM2  R469  00-05-01  RIT-ROOSEVELT        R   
197621        TRAM2  R469  00-05-01  RIT-ROOSEVELT        R   
197622        TRAM2  R469  00-05-01  RIT-ROOSEVELT        R   
197623        TRAM2  R469  00-05-01  RIT-ROOSEVELT        R   
197624        TRAM2  R469  00-05-01  RIT-ROOSEVELT        R   

                                      DIVISION        DATE      TIME     DESC  \
0       Brooklyn–Manhattan Transit Corporation  08/25/2018  00:00:00  REGULAR   
1       Brooklyn–Manhattan Transit Corporation  08/25/2018  04:00:00  REGULAR   
2       Brooklyn–Manhattan Transit Corporation  08/25/2018  08:00:00  REGULAR   
3       Brooklyn–Manhattan Transit Corporation  08/25/2018  12:00:00  REGULAR   
4       Brooklyn–Manhattan Transit Corporation  08/25/2018  16:00:00  REGULAR   
...                                        ...         ...       ...      ...   
197620                   Roosevelt Island Tram  08/31/2018  05:00:00  REGULAR   
197621                   Roosevelt Island Tram  08/31/2018  09:00:00  REGULAR   
197622                   Roosevelt Island Tram  08/31/2018  13:00:00  REGULAR   
197623                   Roosevelt Island Tram  08/31/2018  17:00:00  REGULAR   
197624                   Roosevelt Island Tram  08/31/2018  21:00:00  REGULAR   

           ENTRIES                                              EXITS  \
0       0006736067  0002283184                                    ...   
1       0006736087  0002283188                                    ...   
2       0006736105  0002283229                                    ...   
3       0006736180  0002283314                                    ...   
4       0006736349  0002283384                                    ...   
...            ...                                                ...   
197620  0000005554  0000000348                                    ...   
197621  0000005554  0000000348                                    ...   
197622  0000005554  0000000348                                    ...   
197623  0000005554  0000000348                                    ...   
197624  0000005554  0000000348                                    ...   

        On_N_Line  
0            True  
1            True  
2            True  
3            True  
4            True  
...           ...  
197620      False  
197621      False  
197622      False  
197623      False  
197624      False  

[197625 rows x 12 columns]

Again, note that the dataframe was not automatically transformed by doing this. If we look at it now, `C/A` is still there:

In [ ]:
df

C/A  UNIT       SCP        STATION LINENAME  \
0        A002  R051  02-00-00          59 ST  NQR456W   
1        A002  R051  02-00-00          59 ST  NQR456W   
2        A002  R051  02-00-00          59 ST  NQR456W   
3        A002  R051  02-00-00          59 ST  NQR456W   
4        A002  R051  02-00-00          59 ST  NQR456W   
...       ...   ...       ...            ...      ...   
197620  TRAM2  R469  00-05-01  RIT-ROOSEVELT        R   
197621  TRAM2  R469  00-05-01  RIT-ROOSEVELT        R   
197622  TRAM2  R469  00-05-01  RIT-ROOSEVELT        R   
197623  TRAM2  R469  00-05-01  RIT-ROOSEVELT        R   
197624  TRAM2  R469  00-05-01  RIT-ROOSEVELT        R   

                                      DIVISION        DATE      TIME     DESC  \
0       Brooklyn–Manhattan Transit Corporation  08/25/2018  00:00:00  REGULAR   
1       Brooklyn–Manhattan Transit Corporation  08/25/2018  04:00:00  REGULAR   
2       Brooklyn–Manhattan Transit Corporation  08/25/2018  08:00:00  REGULAR   
3       Brooklyn–Manhattan Transit Corporation  08/25/2018  12:00:00  REGULAR   
4       Brooklyn–Manhattan Transit Corporation  08/25/2018  16:00:00  REGULAR   
...                                        ...         ...       ...      ...   
197620                   Roosevelt Island Tram  08/31/2018  05:00:00  REGULAR   
197621                   Roosevelt Island Tram  08/31/2018  09:00:00  REGULAR   
197622                   Roosevelt Island Tram  08/31/2018  13:00:00  REGULAR   
197623                   Roosevelt Island Tram  08/31/2018  17:00:00  REGULAR   
197624                   Roosevelt Island Tram  08/31/2018  21:00:00  REGULAR   

           ENTRIES                                              EXITS  \
0       0006736067  0002283184                                    ...   
1       0006736087  0002283188                                    ...   
2       0006736105  0002283229                                    ...   
3       0006736180  0002283314                                    ...   
4       0006736349  0002283384                                    ...   
...            ...                                                ...   
197620  0000005554  0000000348                                    ...   
197621  0000005554  0000000348                                    ...   
197622  0000005554  0000000348                                    ...   
197623  0000005554  0000000348                                    ...   
197624  0000005554  0000000348                                    ...   

        On_N_Line  
0            True  
1            True  
2            True  
3            True  
4            True  
...           ...  
197620      False  
197621      False  
197622      False  
197623      False  
197624      False  

[197625 rows x 12 columns]

If we want the change to "stick", one way to do that is to use `inplace=True`:

In [ ]:
df.rename(columns={'C/A' : 'CONTROL_AREA'}, inplace=True)

Now the value has actually been changed:

In [ ]:
df.head()

CONTROL_AREA  ... On_N_Line
0         A002  ...      True
1         A002  ...      True
2         A002  ...      True
3         A002  ...      True
4         A002  ...      True

[5 rows x 12 columns]

Note that this behavior (not changing the contents of the dataframe unless you use `inplace=True` or reassign the variable) is not a mistake or oversight in pandas. It is a useful feature that lets you preview the outcome of an operation before permanently applying it! This is especially important if you are dropping data or transforming it in a way that is not reversible.

### Dropping Columns

Let's say we have determined that the `DESC` column doesn't matter. We can test out dropping it like this:

In [ ]:
df.drop('DESC', axis=1, inplace=True)
df.head()

CONTROL_AREA  ... On_N_Line
0         A002  ...      True
1         A002  ...      True
2         A002  ...      True
3         A002  ...      True
4         A002  ...      True

[5 rows x 11 columns]

Note the `axis=1` argument. By default, `df.drop()` tries to drop rows (`axis=0`) with the specified index, e.g.:

In [ ]:
df.drop(3).head()

CONTROL_AREA  UNIT       SCP STATION LINENAME  \
0         A002  R051  02-00-00   59 ST  NQR456W   
1         A002  R051  02-00-00   59 ST  NQR456W   
2         A002  R051  02-00-00   59 ST  NQR456W   
4         A002  R051  02-00-00   59 ST  NQR456W   
5         A002  R051  02-00-00   59 ST  NQR456W   

                                 DIVISION        DATE      TIME     DESC  \
0  Brooklyn–Manhattan Transit Corporation  08/25/2018  00:00:00  REGULAR   
1  Brooklyn–Manhattan Transit Corporation  08/25/2018  04:00:00  REGULAR   
2  Brooklyn–Manhattan Transit Corporation  08/25/2018  08:00:00  REGULAR   
4  Brooklyn–Manhattan Transit Corporation  08/25/2018  16:00:00  REGULAR   
5  Brooklyn–Manhattan Transit Corporation  08/25/2018  20:00:00  REGULAR   

      ENTRIES                                              EXITS  On_N_Line  
0  0006736067  0002283184                                    ...       True  
1  0006736087  0002283188                                    ...       True  
2  0006736105  0002283229                                    ...       True  
4  0006736349  0002283384                                    ...       True  
5  0006736562  0002283425                                    ...       True

If you are trying to drop a column and you forget the `axis=1`, you'll get an error message like this:

In [ ]:
df.drop('DESC')

Let's go ahead and permanently drop that column:

In [ ]:
df = df.drop('DESC', axis=1) 
df.head()

CONTROL_AREA  UNIT       SCP STATION LINENAME  \
0         A002  R051  02-00-00   59 ST  NQR456W   
1         A002  R051  02-00-00   59 ST  NQR456W   
2         A002  R051  02-00-00   59 ST  NQR456W   
3         A002  R051  02-00-00   59 ST  NQR456W   
4         A002  R051  02-00-00   59 ST  NQR456W   

                                 DIVISION        DATE      TIME     ENTRIES  \
0  Brooklyn–Manhattan Transit Corporation  08/25/2018  00:00:00  0006736067   
1  Brooklyn–Manhattan Transit Corporation  08/25/2018  04:00:00  0006736087   
2  Brooklyn–Manhattan Transit Corporation  08/25/2018  08:00:00  0006736105   
3  Brooklyn–Manhattan Transit Corporation  08/25/2018  12:00:00  0006736180   
4  Brooklyn–Manhattan Transit Corporation  08/25/2018  16:00:00  0006736349   

                                               EXITS  On_N_Line  
0  0002283184                                    ...       True  
1  0002283188                                    ...       True  
2  0002283229                                    ...       True  
3  0002283314                                    ...       True  
4  0002283384                                    ...       True

### Changing Column Types

Another common data munging technique can be reformatting column types. We first previewed column types above using the `df.info()` method, which we'll repeat here.

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 197625 entries, 0 to 197624
Data columns (total 11 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   CONTROL_AREA  197625 non-null  object
 1   UNIT          197625 non-null  object
 2   SCP           197625 non-null  object
 3   STATION       197625 non-null  object
 4   LINENAME      197625 non-null  object
 5   DIVISION      197625 non-null  object
 6   DATE          197625 non-null  object
 7   TIME          197625 non-null  object
 8   ENTRIES       197625 non-null  object
 9   EXITS         197625 non-null  object
 10  On_N_Line     197625 non-null  bool  
dtypes: bool(1), object(10)
memory usage: 15.3+ MB


We can also check the data type of an individual column, rather than listing all of them:

In [ ]:
print(df['ENTRIES'].dtype)

object


In this case we specified `dtype=str` when we opened the file, telling pandas to treat all of the columns as strings initially. So currently every column except for `On_N_Line` is dtype `object`.

A common transformation needed is converting numbers stored as text (dtype `object`) to *float* or *integer* representations.

Let's look more closely at `ENTRIES`:

In [ ]:
df.loc[:5, 'ENTRIES']

0    0006736067
1    0006736087
2    0006736105
3    0006736180
4    0006736349
5    0006736562
Name: ENTRIES, dtype: object

Those seem like integers. Let's try converting the type:

In [ ]:
df.loc[:5, 'ENTRIES'].astype(int)

0    6736067
1    6736087
2    6736105
3    6736180
4    6736349
5    6736562
Name: ENTRIES, dtype: int64

Note that again, we could use `.map()` instead:

In [ ]:
# int is a built-in function, so we do not
# need to declare a helper function here
df.loc[:5, 'ENTRIES'].map(int)

0    6736067
1    6736087
2    6736105
3    6736180
4    6736349
5    6736562
Name: ENTRIES, dtype: int64

That looks good, so let's change the type of that column:

In [ ]:
df['ENTRIES'] = df['ENTRIES'].astype(int)

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 197625 entries, 0 to 197624
Data columns (total 11 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   CONTROL_AREA  197625 non-null  object
 1   UNIT          197625 non-null  object
 2   SCP           197625 non-null  object
 3   STATION       197625 non-null  object
 4   LINENAME      197625 non-null  object
 5   DIVISION      197625 non-null  object
 6   DATE          197625 non-null  object
 7   TIME          197625 non-null  object
 8   ENTRIES       197625 non-null  int64 
 9   EXITS         197625 non-null  object
 10  On_N_Line     197625 non-null  bool  
dtypes: bool(1), int64(1), object(9)
memory usage: 15.3+ MB


Attempting to convert a string column to int or float will produce errors if there are actually non-numeric characters. For example, `LINENAME`:

In [ ]:
df['LINENAME'] = df['LINENAME'].astype(int)

## Converting Dates

A slightly more complicated data type transformation is creating *date* or *datetime* objects. These are pandas data types that have useful information such as being able to quickly calculate the time between two days, or extracting the day of the week from a given date. However, if we look at our current date column, we will notice it is simply a dtype `object` (all strings).

In [ ]:
df['DATE'].head()

0    08/25/2018
1    08/25/2018
2    08/25/2018
3    08/25/2018
4    08/25/2018
Name: DATE, dtype: object

### `pd.to_datetime()`

This is the handiest of methods when converting strings to datetime objects.

Often you can simply pass the series into this function, but it is good practice to preview the results first to prevent overwriting data if some error occurs.

In [ ]:
pd.to_datetime(df['DATE']).head()

0   2018-08-25
1   2018-08-25
2   2018-08-25
3   2018-08-25
4   2018-08-25
Name: DATE, dtype: datetime64[ns]

That worked!

Note that the `dtype` has changed from `object` to `datetime64[ns]`.

Sometimes the above won't work and you'll have to explicitly pass an argument describing how the date is formatted.  
To do that, you have to use some datetime codes. Here's a preview of some of the most common ones:  
<img src="images/strftime_codes.png" width=600>

To explicitly pass formatting parameters, start by previewing your dates to understand their current format as strings.

In [ ]:
# Selecting just the first date entry
df['DATE'].iloc[0] 

'08/25/2018'

Based on that, it looks like we have:

* `08`: a month code with zero padding. So that's `%m` in the table above
* `/`: a delimiter
* `25`: a day of the month. It's not clear that it's zero-padded but we'll go ahead and say it's a `%d` in the table above
* `/`: another delimiter
* `2018`: a year with the century (it would just be `18` without the century). So that's `%Y` in the table above

All together, `%m` + `/` + `%d` + `/` + `%Y` = `%m/%d/%Y`, so we'll use that as the format.

In [ ]:
pd.to_datetime(df['DATE'], format='%m/%d/%Y').head()

0   2018-08-25
1   2018-08-25
2   2018-08-25
3   2018-08-25
4   2018-08-25
Name: DATE, dtype: datetime64[ns]

This has the equivalent behavior for this particular dataset as when we skipped the `format` argument, since pandas was able to detect the format correctly, automatically.

Now let's actually change the whole dataframe's `DATE` to a datetime (skipping the format since we didn't actually need it here):

In [ ]:
df['DATE'] = pd.to_datetime(df['DATE'])
df.head(2)

CONTROL_AREA  UNIT       SCP STATION LINENAME  \
0         A002  R051  02-00-00   59 ST  NQR456W   
1         A002  R051  02-00-00   59 ST  NQR456W   

                                 DIVISION       DATE      TIME  ENTRIES  \
0  Brooklyn–Manhattan Transit Corporation 2018-08-25  00:00:00  6736067   
1  Brooklyn–Manhattan Transit Corporation 2018-08-25  04:00:00  6736087   

                                               EXITS  On_N_Line  
0  0002283184                                    ...       True  
1  0002283188                                    ...       True

In [ ]:
# Make a sample of rows so we can see various dates
date_sample = df['DATE'].sample(n=10, random_state=0)
date_sample

91546    2018-08-30
75472    2018-08-31
151239   2018-08-30
77535    2018-08-25
73591    2018-08-27
10204    2018-08-28
51946    2018-08-27
129569   2018-08-26
10655    2018-08-25
11334    2018-08-30
Name: DATE, dtype: datetime64[ns]

### Applying Datetime Methods

Now that we have converted the `DATE` field to a datetime object we can use some handy built-in methods.

For example, finding the name of the day of the week:

In [ ]:
# .dt stores all the pandas datetime methods (only works for datetime columns)
date_sample.dt.day_name()

91546     Thursday
75472       Friday
151239    Thursday
77535     Saturday
73591       Monday
10204      Tuesday
51946       Monday
129569      Sunday
10655     Saturday
11334     Thursday
Name: DATE, dtype: object

Or, rounding to the nearest 7 days:

In [ ]:
date_sample.dt.round('7D')

91546    2018-08-30
75472    2018-08-30
151239   2018-08-30
77535    2018-08-23
73591    2018-08-30
10204    2018-08-30
51946    2018-08-30
129569   2018-08-23
10655    2018-08-23
11334    2018-08-30
Name: DATE, dtype: datetime64[ns]

## Setting a New Index

It can also be helpful to set one of the columns as the index of the DataFrame, such as when graphing.

In [ ]:
df = df.set_index('DATE')
df.head()

CONTROL_AREA  UNIT       SCP STATION LINENAME  \
DATE                                                       
2018-08-25         A002  R051  02-00-00   59 ST  NQR456W   
2018-08-25         A002  R051  02-00-00   59 ST  NQR456W   
2018-08-25         A002  R051  02-00-00   59 ST  NQR456W   
2018-08-25         A002  R051  02-00-00   59 ST  NQR456W   
2018-08-25         A002  R051  02-00-00   59 ST  NQR456W   

                                          DIVISION      TIME  ENTRIES  \
DATE                                                                    
2018-08-25  Brooklyn–Manhattan Transit Corporation  00:00:00  6736067   
2018-08-25  Brooklyn–Manhattan Transit Corporation  04:00:00  6736087   
2018-08-25  Brooklyn–Manhattan Transit Corporation  08:00:00  6736105   
2018-08-25  Brooklyn–Manhattan Transit Corporation  12:00:00  6736180   
2018-08-25  Brooklyn–Manhattan Transit Corporation  16:00:00  6736349   

                                                        EXITS  On_N_Line  
DATE                                                                      
2018-08-25  0002283184                                    ...       True  
2018-08-25  0002283188                                    ...       True  
2018-08-25  0002283229                                    ...       True  
2018-08-25  0002283314                                    ...       True  
2018-08-25  0002283384                                    ...       True

Or the opposite, resetting the index so that the current index becomes a column and a new index is created:

In [ ]:
df.reset_index()

DATE CONTROL_AREA  UNIT       SCP        STATION LINENAME  \
0      2018-08-25         A002  R051  02-00-00          59 ST  NQR456W   
1      2018-08-25         A002  R051  02-00-00          59 ST  NQR456W   
2      2018-08-25         A002  R051  02-00-00          59 ST  NQR456W   
3      2018-08-25         A002  R051  02-00-00          59 ST  NQR456W   
4      2018-08-25         A002  R051  02-00-00          59 ST  NQR456W   
...           ...          ...   ...       ...            ...      ...   
197620 2018-08-31        TRAM2  R469  00-05-01  RIT-ROOSEVELT        R   
197621 2018-08-31        TRAM2  R469  00-05-01  RIT-ROOSEVELT        R   
197622 2018-08-31        TRAM2  R469  00-05-01  RIT-ROOSEVELT        R   
197623 2018-08-31        TRAM2  R469  00-05-01  RIT-ROOSEVELT        R   
197624 2018-08-31        TRAM2  R469  00-05-01  RIT-ROOSEVELT        R   

                                      DIVISION      TIME  ENTRIES  \
0       Brooklyn–Manhattan Transit Corporation  00:00:00  6736067   
1       Brooklyn–Manhattan Transit Corporation  04:00:00  6736087   
2       Brooklyn–Manhattan Transit Corporation  08:00:00  6736105   
3       Brooklyn–Manhattan Transit Corporation  12:00:00  6736180   
4       Brooklyn–Manhattan Transit Corporation  16:00:00  6736349   
...                                        ...       ...      ...   
197620                   Roosevelt Island Tram  05:00:00     5554   
197621                   Roosevelt Island Tram  09:00:00     5554   
197622                   Roosevelt Island Tram  13:00:00     5554   
197623                   Roosevelt Island Tram  17:00:00     5554   
197624                   Roosevelt Island Tram  21:00:00     5554   

                                                    EXITS  On_N_Line  
0       0002283184                                    ...       True  
1       0002283188                                    ...       True  
2       0002283229                                    ...       True  
3       0002283314                                    ...       True  
4       0002283384                                    ...       True  
...                                                   ...        ...  
197620  0000000348                                    ...      False  
197621  0000000348                                    ...      False  
197622  0000000348                                    ...      False  
197623  0000000348                                    ...      False  
197624  0000000348                                    ...      False  

[197625 rows x 11 columns]

## Summary

We've seen in this lesson the differences between Pandas (Series and DataFrames) and base Python (Dictionaries and Lists) data types. Then we walked through transforming the values in a pandas Series, modifying the columns of a pandas DataFrame, and finally modifying the DataFrame index.

# Statistical Methods in Pandas

## Introduction

In this lesson, you'll learn how to use some of the key summary statistics methods in Pandas.

## Objectives:

You will be able to:

- Calculate summary statistics for a series and DataFrame 
- Use the `.apply()` or `.applymap()` methods to apply a function to a pandas series or DataFrame  


## Getting DataFrame-Level Summary Statistics

When working with a new dataset, the first step is always to begin to understand what makes up that dataset. The Pandas DataFrame class contains two built-in methods that make this very easy for us. 

### Using `df.info()`

The `df.info()` method provides us with summary **_metadata_** about our DataFrame -- that is, it gives us data about our dataset, such as how many rows and columns it contains, and what data types they are stored as.  Let's demonstrate this by reading in the Titanic dataset and calling the `.info()` method on the DataFrame. 

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import pandas as pd
df = pd.read_csv('/content/drive/MyDrive/DATA/titanic.csv', index_col=0)
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    object 
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(4), object(6)
memory usage: 90.5+ KB


As we can see from the output above, the `.info()` method provides us with great information about the characteristics of the DataFrame, without telling us anything about the data it actually contains. 

Examine the output above, and take note of the important things it tells us about the DataFrame, such as:

* The number of columns and rows in the DataFrame
* The data type of the data each column contains
* How many values each column contains (NaNs are not counted)
* The memory footprint of the DataFrame

This sort of information about a dataset is called **_metadata_**, since it's data about our data. 


### Using `.describe()` 

The next step in Exploratory Data Analysis (EDA) is usually to dig into the summary statistics of the dataset, and get a feel for the data each column contains.  Rather than force us to deal with the tedium of doing this individually for every column, Pandas DataFrames provide the handy `df.describe()` method which calculates the basic summary statistics for each column for us automatically. 

See the example in the cell below.

In [ ]:
df.describe()

PassengerId    Survived         Age       SibSp       Parch        Fare
count   891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean    446.000000    0.383838   29.699118    0.523008    0.381594   32.204208
std     257.353842    0.486592   14.526497    1.102743    0.806057   49.693429
min       1.000000    0.000000    0.420000    0.000000    0.000000    0.000000
25%     223.500000    0.000000   20.125000    0.000000    0.000000    7.910400
50%     446.000000    0.000000   28.000000    0.000000    0.000000   14.454200
75%     668.500000    1.000000   38.000000    1.000000    0.000000   31.000000
max     891.000000    1.000000   80.000000    8.000000    6.000000  512.329200

As we can see, the output of the `.describe()` method is very handy, and gives us relevant information such as:

* a `count` of the number of values in each column, making it identify columns with missing values
* The mean and standard deviation of each column
* The minimum and maximum values found in each column
* The median (50%) and quartile values (25% & 75%) for each column

Use the `.describe()` method to quickly help you get a feel for your datasets when you start the Exploratory Data Analysis process. 


## Calculating Individual Column Statistics


If we need to calculate individual statistics about a column, we can also do this easily.  Pandas DataFrames and Series objects come with a plethora of built-in methods to instantly calculate summary statistics for us. 

See the code blocks below for examples:

In [ ]:
# Get the mean of every numeric column at once
df.mean()

PassengerId    446.000000
Survived         0.383838
Age             29.699118
SibSp            0.523008
Parch            0.381594
Fare            32.204208
dtype: float64

In [ ]:
# Get the mean of a specific column
df['Fare'].mean()

32.2042079685746

In [ ]:
# Get the value for 90% quantile for a specific column
df['Age'].quantile(.9)

50.0

In [ ]:
# Get the median value for a specific column
df['Age'].median()

28.0

There are many different statistical methods built into Pandas DataFrames -- these are just a few! We will not list all of them, but here are some common ones you'll probably make use of early and often:

* `.mode()` -- the mode of the column
* `.count()` -- the count of the total number of entries in a column
* `.std()` --  the standard deviation for the column
* `.var()` -- the variance for the column  
* `.sum()` -- the sum of all values in the column
* `.cumsum()` -- the cumulative sum, where each cell index contains the sum of all indices lower than, and including, itself.


### Summary Statistics for Categorical Columns

Obviously, we cannot calculate most summary statistics on columns that contain non-numeric data -- there's no way for us to find the mean of the letters in the `Embarked` column, for instance.  However, there are some summary statistics we can use to help us better understand our categorical columns. 

See the examples in the cell below:

In [ ]:
df['Embarked'].unique()

array(['S', 'C', 'Q', nan], dtype=object)

In [ ]:
df['Embarked'].value_counts()

S    644
C    168
Q     77
Name: Embarked, dtype: int64

These methods are extremely useful when dealing with categorical data! 

`.unique()` shows us all the unique values contained in the column. 

`.value_counts()` shows us a count for how many times each unique value is present in a dataset, giving us a feel for the distribution of values in the column. 

### Calculating on the Fly with `.apply()` and `.applymap()`

Sometimes, we'll need to make changes to our dataset, or to compute functions on our data that aren't built-in to Pandas.  We can do this by passing lambda values into the `apply()` method when working with Pandas series, and the `.applymap()` method when working with Pandas DataFrames. 

Note that both of these do not mutate the original dataset -- instead, they return a copy of the Series or DataFrame containing the result. 

See the example in the cell below:

In [ ]:
# Quick function to convert every value in the DataFrame to a string
string_df = df.applymap(lambda x: str(x))
string_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   PassengerId  891 non-null    object
 1   Survived     891 non-null    object
 2   Pclass       891 non-null    object
 3   Name         891 non-null    object
 4   Sex          891 non-null    object
 5   Age          891 non-null    object
 6   SibSp        891 non-null    object
 7   Parch        891 non-null    object
 8   Ticket       891 non-null    object
 9   Fare         891 non-null    object
 10  Cabin        891 non-null    object
 11  Embarked     891 non-null    object
dtypes: object(12)
memory usage: 90.5+ KB


In [ ]:
# Let's quickly square every value in the Age column
display(df['Age'].apply(lambda x: x**2).head())

# Note that the original data in the age column has not changed
df['Age'].head()

df['AgeSq'] = df['Age'].apply(lambda x: x**2)
df.head()

0     484.0
1    1444.0
2     676.0
3    1225.0
4    1225.0
Name: Age, dtype: float64

PassengerId  Survived Pclass  ... Cabin Embarked   AgeSq
0            1         0      3  ...   NaN        S   484.0
1            2         1      1  ...   C85        C  1444.0
2            3         1      3  ...   NaN        S   676.0
3            4         1      1  ...  C123        S  1225.0
4            5         0      3  ...   NaN        S  1225.0

[5 rows x 13 columns]

## Summary

In this lesson, you learned how to:

* Understand and use the `df.describe()` and `df.info()` summary statistics methods 
* Use built-in Pandas methods for calculating summary statistics 
* Apply a function to every element in a Series or DataFrame using `s.apply()` and `df.applymap()` 

In [ ]:
import pandas as pd
legos = pd.read_csv('/content/drive/MyDrive/DATA/lego_setsHB.csv')
legos.head()

list_price  num_reviews  ...  star_rating  val_star_rating
0       29.99          2.0  ...          4.5              4.0
1       19.99          2.0  ...          5.0              4.0
2       12.99         11.0  ...          4.3              4.1
3       99.99         23.0  ...          4.6              4.3
4       79.99         14.0  ...          4.6              4.1

[5 rows x 8 columns]

1. use the .info() and .describe() functions to have an overall description of the dataset. 

2. use boolean operators find the Lego sets which have star_rating greater than or equal to 4.

3. find the number of Lego sets whose star_rating greater than or equal to 4. (Hint: conside using .count( ) on the specified column)

4. find the average list price for the Lego sets whose star_rating greater than or equal to 4. (Hint: consider specifying the list_price column after obtaining data and before taking the average)

In [ ]:
legos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12261 entries, 0 to 12260
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   list_price         12261 non-null  float64
 1   num_reviews        10641 non-null  float64
 2   piece_count        12261 non-null  int64  
 3   play_star_rating   10486 non-null  float64
 4   prod_id            12261 non-null  int64  
 5   review_difficulty  10206 non-null  object 
 6   star_rating        10641 non-null  float64
 7   val_star_rating    10466 non-null  float64
dtypes: float64(5), int64(2), object(1)
memory usage: 766.4+ KB


In [ ]:
legos.describe()

list_price   num_reviews  ...   star_rating  val_star_rating
count  12261.000000  10641.000000  ...  10641.000000     10466.000000
mean      65.141998     16.826238  ...      4.514134         4.228960
std       91.980429     36.368984  ...      0.518865         0.660282
min        2.272400      1.000000  ...      1.800000         1.000000
25%       19.990000      2.000000  ...      4.300000         4.000000
50%       36.587800      6.000000  ...      4.700000         4.300000
75%       70.192200     13.000000  ...      5.000000         4.700000
max     1104.870000    367.000000  ...      5.000000         5.000000

[8 rows x 7 columns]

In [ ]:
legos[legos['star_rating']>=4]

list_price  num_reviews  ...  star_rating  val_star_rating
0         29.9900          2.0  ...          4.5              4.0
1         19.9900          2.0  ...          5.0              4.0
2         12.9900         11.0  ...          4.3              4.1
3         99.9900         23.0  ...          4.6              4.3
4         79.9900         14.0  ...          4.6              4.1
...           ...          ...  ...          ...              ...
12255     36.5878          8.0  ...          4.6              4.5
12256     36.5878          6.0  ...          4.3              4.2
12258     24.3878         18.0  ...          4.6              4.5
12259     12.1878          1.0  ...          5.0              5.0
12260     12.1878         11.0  ...          4.7              4.8

[9563 rows x 8 columns]

In [ ]:
legos[legos['star_rating']>=4].count()

list_price           9563
num_reviews          9563
piece_count          9563
play_star_rating     9408
prod_id              9563
review_difficulty    9128
star_rating          9563
val_star_rating      9388
dtype: int64

In [ ]:
legos4 = legos[legos['star_rating']>=4]
legos4.shape

(9563, 8)

In [ ]:
legos.loc[legos['star_rating']>=4, ['list_price']].mean()

list_price    68.24627
dtype: float64

In [ ]:
legos4['list_price'].mean()

68.24627018717933

In [ ]:
legos.iloc[1,3]
legos.loc[1, 'list_price']

19.99